vega/altair.py

```diff
@@ -0,0 +1,155 @@
+# Functions to stream Altair specifications
+
+import time
+import warnings
+
+from toolz import curried
+
+from IPython.display import display
+
+import pandas as pd
+import altair as alt
+from altair import examples
+from altair.utils.execeval import eval_block
+
+from .widget import VegaWidget
+
+
+def _exceptions(exceptions):
+    if exceptions is None:
+        exceptions = {}
+    elif isinstance(exceptions, list):
+        exceptions = set([id(exc) if isinstance(exc, pd.DataFrame) else exc
+                          for exc in exceptions])
+    return exceptions
+
+
+@curried.curry
+def to_streaming(data, context=None, exceptions=None, debug=False):
+    id_ = id(data)
+    if debug:
+        print('to_streaming '+str(id_))
+    exceptions = _exceptions(exceptions)
+    if not isinstance(data, pd.DataFrame):
+        raise TypeError(f"Expected DataFrame got: {type(data)}")
+    if hasattr(data, "__geo_interface__"):
+        # raise TypeError("Unhandled __geo_interface__ for now")
+        warnings.warn("Unhandled __geo_interface__, not streaming")
+        return alt.to_values(data)
+    if id_ in exceptions:
+        if debug:
+            print('Exception for', id_)
+        return alt.to_values(data)
+    if isinstance(context, dict):
+        name = context.get(id_, None)
+        if name is None:
+            if len(context) == 0:
+                name = "data"
+            else:
+                name = f"data-{len(context)+1}"
+            context[id_] = (name, data)
+        else:
+            name = name[0]
+    else:
+        name = "data"
+    return {"name": name}
+
+
+alt.data_transformers.register('streaming', to_streaming)
+
+
+def stream(alt_spec, exceptions=None, reuse=False, resize=True, debug=False):
+    """
+    Send an Altair specification to the notebook using the streaming API
+
+    Parameters
+    ----------
+    alt_spect : top-level chart object such as Chart or LayeredChart
+        the Altair specification
+    exceptions : list
+        a list of dataframes of id(dataframe) that should be inlined
+    reuse : boolean
+        if True, return the context dictionary and the widget. The
+        context dictionary associate a dataframe id with a pair
+        (name, data). The name is useful to reuse the widget for
+        streaming more data using the widget `.update` method.
+    resize: boolean
+        Resize vega component after sending the update. Some charts
+        need it, others don't. Defaults to True but can be overridden
+        here.
```

Comment:
Vega and Vega-Lite have an autosize property that we could use instead of defining it here: https://vega.github.io/vega-lite/docs/spec.html#top-level.

Reply:
Addressed. We need to manually resize.